### ISSUE TITLE
"Connection reset by peer" if container is used to quickly after creation

### ISSUE THREAD

**Issue by ghost** [testcontainers/testcontainers-rs#752]
Hello,

I am writing a test that creates a PostgreSQL container and then uses `sqlx` to connect to it. The goal is to test the SQL migrations.

```rust
#[cfg(test)]
mod tests {
    use sqlx::postgres::PgPoolOptions;
    use std::process::Command;
    use testcontainers::{
        core::{IntoContainerPort, WaitFor},
        runners::AsyncRunner,
        GenericImage, ImageExt,
    };

    #[tokio::test]
    async fn ensure_database_migration_is_valid() {
        let postgres_container = GenericImage::new("postgres", "14")
            .with_exposed_port(5432.tcp())
            .with_wait_for(WaitFor::message_on_stdout(
                "database system is ready to accept connections",
            ))
            // .with_wait_for(WaitFor::seconds(1)) // WORK AROUND, SEE TEXT BELOW
            .with_env_var("POSTGRES_HOST_AUTH_METHOD", "trust")
            .start()
            .await
            .unwrap();

        let output = Command::new("netstat").arg("-tunlp").output().unwrap();
        println!("status: {}", output.status);
        println!("stdout: {}", String::from_utf8_lossy(&output.stdout));
        println!("stderr: {}", String::from_utf8_lossy(&output.stderr));

        let output = Command::new("docker").arg("ps").output().unwrap();
        println!("status: {}", output.status);
        println!("stdout: {}", String::from_utf8_lossy(&output.stdout));
        println!("stderr: {}", String::from_utf8_lossy(&output.stderr));

        let postgres_host = postgres_container.get_host().await.unwrap();
        let postgres_host_port = postgres_container.get_host_port_ipv4(5432).await.unwrap();
        let database_url =
            format!("postgres://postgres@{postgres_host}:{postgres_host_port}/postgres");
        println!("{}", database_url);

        let pool = PgPoolOptions::new()
            .max_connections(5)
            .connect(&database_url)
            .await
            .unwrap();
        sqlx::migrate!("db/migrations").run(&pool).await.unwrap();
    }
}
```

Sometimes, especially when the tests run in a Bitbucket pipeline, the test fails as follows.

```
---- tests::ensure_database_migration_is_valid stdout ----
status: exit status: 0
stdout: Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:29418           0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:32768           0.0.0.0:*               LISTEN      -
tcp6       0      0 :::32768                :::*                    LISTEN      -
tcp6       0      0 :::2375                 :::*                    LISTEN      -

stderr:
status: exit status: 0
stdout: CONTAINER ID   IMAGE         COMMAND                  CREATED        STATUS                  PORTS                                         NAMES
3265d42a073e   postgres:14   "docker-entrypoint.s..."   1 second ago   Up Less than a second   0.0.0.0:32768->5432/tcp, :::32768->5432/tcp   zealous_hofstadter

stderr:
postgres://postgres@localhost:32768/postgres
thread 'tests::ensure_database_migration_is_valid' panicked at src/main.rs:86:14:
called `Result::unwrap()` on an `Err` value: Io(Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" })


failures:
    tests::ensure_database_migration_is_valid

test result: FAILED. 3 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 8.35s

error: test failed, to rerun pass `--bin backend`
error: 1 target failed:
    `--bin backend`
```

As you can see, I already have some debug output in there (`netstat -tunlp`, `docker ps`, print the resulting database connection string). The problem is that the container or Dockers network setup or something else is not ready before `sqlx` tries to connect to PostgreSQL.

Work around: our work around is to add the `.with_wait_for(WaitFor::seconds(1))`, which is commented in the code listing above.

Now, in the end, this may be a problem with `sqlx`, which should be more robust, when connecting to the database. No questions asked.

On the other hand, would it be possible to work around this problem in `testcontainers-rs`? I am thinking about something like `.with_wait_for(WaitFor::exposed_ports_accept_connections())` (which does not exist).

Is this in the scope of `testcontainers-rs` at all? Has anybody else experienced this problem?

**Comment by DDtKey** [comment]
Hi @baghai 👋 

Have you tried ready-to-use modules? I see that you have only one `WaitFor` condition, but we have expirienced similar issue due to this: [fix](https://github.com/testcontainers/testcontainers-rs-modules-community/pull/162) and [the issue description](https://github.com/testcontainers/testcontainers-rs-modules-community/issues/158)

TLDR: postgres runs init-script and then restarts, so it's not enough to wait for log message only 1 time. This may help in your case.
